### ISSUE TITLE
Week 3 - APIs (cohort feedback)

### ISSUE THREAD

**Issue by mantagen** [foundersandcoders/master-reference#630]
[Week schedule](https://github.com/foundersandcoders/master-reference/tree/master/coursebook/week-3)

Week facilitator: @mantagen 
FACG2: @Walidmash @Eslam3bed 
FACN2: @Heathercoraje
FAC11: @rebecacalvoquintero @ameliejyc 

#### SGCs
+ [FAC11 students](https://github.com/foundersandcoders/london-programme/blob/master/stop-go-continue/fac-11/cohort-week-3.md)
+ [FAC11 mentors](https://github.com/foundersandcoders/master-reference/blob/master/curriculum-planning/minutes/api/summer-2017-handover-london-naz.md)
+ [FACN2 students](https://github.com/foundersandcoders/nazareth-programme/blob/master/stop-go-continue/facn2/week3.md)
+ [FACG2 students](https://github.com/foundersandcoders/gaza-programme/blob/master/stop-go-continue/FACG2/)

#### Open issues that relate to your week
[Week 3 issues](https://github.com/foundersandcoders/master-reference/issues?q=is%3Aopen+is%3Aissue+label%3Aweek-apis)

**Comment by Eslam3bed** [comment]
## problem 2:
- Full day just for Css and FlexBox to practice flex more and  get familiar with it - in day 2 [ref](https://github.com/foundersandcoders/master-reference/tree/master/coursebook/week-3#day-2)

**Comment by Walidmash** [comment]
Problem 1:
In hxr wrokshop, we can:
- add ajax introduction.
- simple example about ajax.
- make a comparison between ajax and xhr. 
- when to use each.
students need to know that there is a technology called ajax and how to works, so they can know when to use it and how to handle it if they have to edit a website based on ajax.

**Comment by Heathercoraje** [comment]
Reviewer: @veredrec @heathercoraje
#### Introductory Worshop
We think this introductory workshop contains important topics. Among topics, XHLHttpRequest is very well-explained and also `line by line` part is very helpful. We suggest following the style of XHLHttprequest as in to have more intuitive example. 

#### XHR Workshop
We think the content of workshop is good. However, @heathercoraje thinks that the workshop needs a bit more time or better instruction leading mentors as I remember many of the cohort struggle to get started.

#### Workshop APIs
Great workshop

#### Flexbox Froggy Workshop
One of the best workshops
Our only suggestion is that this workshop can be moved to afternoon when students suffer food comma.

#### Flexbox Dice Morning Challenge
Great workshop. It is not too difficult but not to easy. It is a good workshop to move forward from basic flexbox to a bit more advanced level.

#### Software Design Workshop
We think the topic itself, architecture, is important but it will be much better to go through the drawing process of architecture or draw diagrams and file structure in pairs instead of moving files in file tree.
The workshop has lots of text worth reading but the task at the end of repo could be improved to actually implement or practice the architecture process in the workshop. 

#### ws-software-design-js
We think this needs to be separated into different workshops or break it down so that cohort completes first topic then we take 15 mins to go over solution. @heathercoraje remember this workshop quite heavy and difficult to understand the aim itself. @veredrec thinks it is more a matter of large amount of materials in a single workshop

#### Parallel function 
As callbacks are very important, we think that it is better to start with a very simple example of callbacks in order to introduce the importance/ utility of callbacks.
@heathercoraje I think structure of parallel function didn't deliver the concept of callbacks.

**Comment by ameliejyc** [comment]
## London
#### Reviewers: @ameliejyc and @rebecacalvoquintero 

### Introductory workshop (API basics, HTTP, XMLHttp requests, request-response pattern)
* We think this workshop can benefit from a short introductory presentation to get students started.
* The reading material is quite long and should be broken up by mentors asking questions and reviewing the material. 
* Perhaps more information on GET and POST requests given we use these heavily in later projects.

### XHR workshop (Giphy)
* We think this was a fun workshop (gifs are great). Given we're introducing students to Postman, a few minutes should be spent explaining this.

### Afternoon workshop (making API requests in the browser, callbacks, working with JSON, chaining API requests)
* Good workshop and choice of API to work with. However we're not sure how far students got into the coverage/testing section and this would surely also require a bit more explanation. 

### Flexbox Froggy
* Amazing website and challenge. It makes a nice visual change to work on a fun website like this as most morning challenges are run locally.

### Flexbox Dice Morning Challenge
* Also brilliant challenge. Leads on perfectly from Flexbox Froggy.

### Software Architecture Workshop
* A _lot_ of reading. We think this workshop could also benefit from an introductory presentation.
* We also feel this workshop needs to include more interactivity. We could ask students to apply what they learn during this workshop to a previous project to work out how they would have architected it.
* Drawing out software paths might be a better task than moving files around, especially as software architecting on whiteboards is encouraged in later projects.

### Software Design Workshop
* Another large amount of reading. We would like to have distinct breaks between the 3 topics in Stage 1 (Functions as First Class Objects, Closures and Scope, Abstraction with Functions) where we lollipop questions and discuss the material briefly.

### Parallel function morning challenge
* Our cohort struggled with this challenge. The challenge should be explained in more depth and we should budget extra time afterwards to go through it as most people required it. Difficult challenges are good but we should ensure that we check in with the students midway through and offer pointers so as not to demoralise anyone if they are struggling.

**Comment by mantagen** [comment]
# Week 3 (API) Curriculum Planning Meeting 20/09/2017

## Intended outcomes
- List suggested changes
- List action points for each change
- Record decision making process

## Meta

### Facilitator
Matt G (@mantagen)
### Notetaker
Amelie (@ameliejyc)


## Sections
### WS - Intro
> #### Gaza
> - In the HTTP part,@Eslam3bed thinks that the description was abstracte and the differance between the http methods are confusing , especially the difference between POST and PUT.
> - @Walidmash: the tech community is using the concept of "Ajax", and we know nothing about it, even though I think we knew some about it indirectly, Its related to hxr workshop in this week, as for suggestions,in hxr workshop we can:
>   - add ajax introduction.
>   - add simple example about ajax.
>   - explain relation between ajax and xhr.
when and how to use it.
>- I think students need to know that there is a technology called ajax and how it works, so they can know when to use it and how to handle it if they have to edit a website based on ajax.
>


> #### Nazareth
> - We think this introductory workshop contains important topics. Among topics, XHLHttpRequest is very well-explained and also line by line part is very helpful. We suggest following the style of XHLHttprequest as in to have more intuitive example.

> #### London
> - We think this workshop can benefit from a short introductory presentation to get students started.
> - The reading material is quite long and should be broken up by mentors asking questions and reviewing the material.
> - Perhaps more information on GET and POST requests given we use these heavily in later projects.

#### Minutes


- Gaza: workshop was a bit confusing to understand. Comparison between get and post and this is not clear. Differences betweeen each one? when do we use each of them? Examples would help to make this more clear.
- After week3, make an api call was a constant in every project. That is why there should be more examples for both methods
- having a learning outcomes before each section of reading material.
- a brief introduction to topics before the workshop (this would help studdents to break up the long reading.)    
- Reading broken up (section before the reading with the learning outcomes, review happens, and then people move on at the same time)
- asking questions after a small period of time.
- AP: Add examples to help differentiate between GET/POST methods.
- AP: Make the difference more clear.
- AP: include an introduction to AJAX( because it is not introduced);
- AP: make sure that xhr request can handle redirect responses.
- AP: Brief introduction at the beginning of the workshop --> including the useful learning outcomes the students will get from the workshop.
- AP: Mentors should go through each section with the group at some time intervals.

### WS - XHR
> #### Nazareth
> - We think the content of workshop is good. However, @Heathercoraje thinks that the workshop needs a bit more time or better instruction leading mentors as I remember many of the cohort struggle to get started.

> #### London
> - We think this was a fun workshop (gifs are great). Given we're introducing students to Postman, a few minutes should be spent explaining this.

#### Minutes

- isntallation help? lINUX Students had problems and didnt get much from the Postman workshop
- students didnt know what to do, the workshop was not finished,
- instructions poor on the XHR Workshop 

- AP: Brief introduction to Postman - quick talk through and a quick screencast of a url and what it looks like -  (what it does, what it's used for... );

- AP: More detailed intructions for the XHR workshop: explaining what the students should get from this workshop and checking up they understand the exercise.
> 

### WS - API
> #### Gaza
> - In the code coverage ,@Eslam3bed for me I Didn't understande that part and It was not clear ,@Walidmash there are websites like Code Climate uses istanbul for code coverage, its very helpful in projects, it would be useful if students knew about it and the use of it.

> #### Nazareth
> - Great workshop

> ### London
> - Good workshop and choice of API to work with. However we're not sure how far students got into the coverage/testing section and this would surely also require a bit more explanation.
> 

#### Minutes

- London: our cohort didnt finish this workshop. 
- Nazareth: They did some code coverage, but not enough --> 5/10 min. It could be introduced in another workshop instead of adding this as another challenge.
- Maybe code coverage in TDD week? 
- AP: Spend more time and more explanation on code coverage
- AP: Longer workshop or rebalance the different sections of the workshop.
- AP: TDD week should have a workshop/morning challenge regarding code coverage

### MC - Flexbox Froggy
> #### Gaza
> - it was great workshop @Eslam3bed we need more time for advanced Css topics like transitions and FlexBox to practice it more and get familiar with it.
> -as for the morning challege most of students didn't completed it.

> #### Nazareth
> - One of the best workshops
Our only suggestion is that this workshop can be moved to afternoon when students suffer food comma.

> #### London
> - Amazing website and challenge. It makes a nice visual change to work on a fun website like this as most morning challenges are run locally.
> 

#### Minutes

- Nazareth: It should be moved to the afternoon, as students prefer more interactive workshops after lunch.
- London: Flexbox and then dice challenge as they are more related
- Matt: in the following morning, dice challenge would also be ok as it would still be fresh.
- AP: Flexbox in the afternoon, parallel functions in the morning - first thing.
- AP: Matt will remind the rest of the mentors to have a CSS challenge every week.


### MC - Flexbox Dice
> #### Gaza
> - most of students didn't completed it.

> #### Nazareth
> - Great workshop. It is not too difficult but not to easy. It is a good workshop to move forward from basic flexbox to a bit more advanced level.

> #### London
> - Also brilliant challenge. Leads on perfectly from Flexbox Froggy.


#### Minutes

- Gaza: Many students didnt finish it as well as the Flexbox froggy challenge.
- Matt: Mentors to check on the students--> if they have problems and help them when needed.


### WS - Software Architecture
> #### Gaza
> - the workshop was postpond to the end of the week and it was self-study which was not efficient, we suggest that one of the mentors summarise it for the student to save time.

> #### Nazareth
> - We think the topic itself, architecture, is important but it will be much better to go through the drawing process of architecture or draw diagrams and file structure in pairs instead of moving files in file tree.
> -The workshop has lots of text worth reading but the task at the end of repo could be improved to actually implement or practice the architecture process in the workshop.

> #### London
> - A lot of reading. We think this workshop could also benefit from an introductory presentation.
> - We also feel this workshop needs to include more interactivity. We could ask students to apply what they learn during this workshop to a previous project to work out how they would have architected it.
> - Drawing out software paths might be a better task than moving files around, especially as software architecting on whiteboards is encouraged in later projects.
> 
Minutes:

- Gaza: confusing workshop, too long, students didnt understand it. Mentors should explain it --> students would benefit from this.

- London: Lots to read. Short presentation would help students to understand the reading. Workshop should be more interactive --> whiteboard software paths as we will implement this in our future projects. 
- Nazareth: Lots of text. Task was weird --> moving files didnt make that much sense. An explanation re file structure before the task would have been more helpful than just telling the students to move files around.
- Nazareth: Would the SA make more sense to do it on week 4/node week1? Maybe in week3 and in week4? 
- london: it is confusing when you add the backend. But it should be introduced on week 3 too.
- AP: This workshop should be more interactive. (Moving files around--> not enough). Worskhop to be refactored.

- AP: TOP up of software architecture in the following weeks --> mentors should encourage students in the following week to think about SA.

### WS - Software Design Workshop
> #### Nazareth
> - We think this needs to be separated into different workshops or break it down so that cohort completes first topic then we take 15 mins to go over solution. @Heathercoraje remember this workshop quite heavy and difficult to understand the aim itself. @veredrec thinks it is more a matter of large amount of materials in a single workshop

> #### London
> - Another large amount of reading. We would like to have distinct breaks between the 3 topics in Stage 1 (Functions as First Class Objects, Closures and Scope, Abstraction with Functions) where we lollipop questions and discuss the material briefly.
> 

#### Minutes

> Gaza: no comments
> Nazareth: Students didnt feel the importance of the projects, they didnt get much from the workshop. Workshop should be broken down in pieces.
> London: Group discussion? Lots of reading material again. It should be broken down: questions, learning outcomes at the begginning, make sure people understand the workshop before moving on.
> 
- AP: Breaks with questions, explanations and discussion re the solution


### MC - Parallel function
> #### Gaza
> - @Eslam3bed, we did finish this morning challenge, but we didn't use it in any project, maybe because we dont know when we have to use it and its value.

> #### Nazareth
> - As callbacks are very important, we think that it is better to start with a very simple example of callbacks in order to introduce the importance/ utility of callbacks.
> - @Heathercoraje I think structure of parallel function didn't deliver the concept of callbacks.

> #### London
> - Our cohort struggled with this challenge. The challenge should be explained in more depth and we should budget extra time afterwards to go through it as most people required it. Difficult challenges are good but we should ensure that we check in with the students midway through and offer pointers so as not to demoralise anyone if they are struggling.
> 

#### Minutes

- Nazareth: Everybody was lost, no direction. It took much longer. And understanding the solution took ages as well because of this lack of direction/understanding re callbacks. Structure of parallel functions was too compliated. Biggest problem how the workshop and the function were formatted.
- London: Lack of understanding on callbacks --> people struggled for ages without any direction. Explanation was longer as callbacks were not really well explained beforehand. We had been introduced to callbacks before - traffic light challenge - the way the function was laid out it was very confusion --> better introduction to this type of function would have been very useful.
- Gaza: They already knew callbacks. The problems was they didnt understand how they would implement parallel/waterfall functions. They had the traffic light challenge before. More simple parallel functions.
- London: We should not put too much weight on the mentor --> it could be explained through a diagram/youtube video.

- AP: Better explanation of the structure, value, implementation of parallel functions.

- AP: A visual representation of parallel functions would be very useful

- AP: Find the best youtube video on parallel function, less than 5 mins. Link them to the best part of the youtube video.

**Comment by mantagen** [comment]
# Action Points from meeting 20/09/2017
## Sections
### WS - Intro
- AP: Add examples to help differentiate between GET/POST methods.
- AP: Make the difference more clear.
- AP: include an introduction to AJAX( because it is not introduced);
- AP: make sure that xhr request can handle redirect responses.
- AP: Brief introduction at the beginning of the workshop --> including the useful learning outcomes the students will get from the workshop.
- AP: Mentors should go through each section with the group at some time intervals.

### WS - XHR
- AP: Brief introduction to Postman - quick talk through and a quick screencast of a url and what it looks like -  (what it does, what it's used for... );
- AP: More detailed intructions for the XHR workshop: explaining what the students should get from this workshop and checking up they understand the exercise.

### WS - API
- AP: Spend more time and more explanation on code coverage
- AP: Longer workshop or rebalance the different sections of the workshop.
- AP: TDD week should have a workshop/morning challenge regarding code coverage

### MC - Flexbox Froggy
- AP: Flexbox in the afternoon, parallel functions in the morning - first thing.
- AP: Matt will remind the rest of the mentors to have a CSS challenge every week.


### MC - Flexbox Dice
- AP: Mentors to ask students to revise Flexbox Froggy if necessary in the gap between Flexbox Froggy and Flexbox Dice.

### WS - Software Architecture
- AP: This workshop should be more interactive. (Moving files around--> not enough). Worskhop to be refactored.
- AP: Top up of software architecture in the following weeks --> mentors should encourage students in the following week to think about SA.

### WS - Software Design Workshop
- AP: Breaks with questions, explanations and discussion re the solution

### MC - Parallel function
- AP: Better explanation of the structure, value, implementation of parallel functions.
- AP: A visual representation of parallel functions would be very useful
- AP: Find the best youtube video on parallel function, less than 5 mins. Link them to the best part of the youtube video.
